esbuild: new experimental option for web builds

sqs opened the pull request:
[esbuild](https://esbuild.github.io/) is an alternative to Webpack for building `client/web` (the web app). Its usage on our codebase is **EXPERIMENTAL** and optional.

To use esbuild instead of Webpack, set the env var `DEV_WEB_BUILDER=esbuild`:

``` shell
DEV_WEB_BUILDER=esbuild enterprise/dev/start.sh
```

Comparison vs. Webpack:

- esbuild: faster initial build (esbuild ~3s vs. Webpack ~53s)
- esbuild: faster recompilation (esbuild ~900ms vs. Webpack ~5000ms)
- esbuild: smaller total asset size in dev (Chrome devtools network resources size for `/search`: esbuild ~24.1MB vs. Webpack ~64.1MB)
- esbuild: faster DOMContentLoaded (on `/search`: esbuild ~1.2s vs. Webpack ~2.3s)
- Webpack: [fast refresh](https://www.npmjs.com/package/react-refresh) (not supported/implemented yet in esbuild, so you need to manually reload the page after each change)

Notes:

- It's probably possible to configure Webpack to be faster and produce smaller dev bundles, so consider these comparisons as reflecting the current state, not the hypothetical ideal state after more optimization.
- Webpack is still used for all other web builds (including storybooks and the browser extension).
- esbuild is not configured to make a production build. Just use it for local dev for now.

Questions or problems with esbuild? Ask in [#frontend-platform](https://app.slack.com/client/T02FSM7DL/C01LTKUHRL3) and mention `@sqs` (who is responsible for this esbuild experiment).

sqs commented:
@valerybugakov This is almost all new/separate code, not code that touches our current Webpack build. The parts that touch our current Webpack build are:

- Factoring out options for the monaco-editor Webpack plugin
- Gulp commands (switching based on the value of `DEV_WEB_BUILDER`)
- `app.html` tags for `<script>` and `<link rel="stylesheet">`

I tested this branch and confirmed all of the following build variants work:

- with Webpack in dev mode (using the Webpack dev server)
- with Webpack in a production build
- with esbuild in dev mode

valerybugakov reviewed (approved):
Thanks for bringing the `esbuild` experiment to life! The speed gains are impressive. It's always good to have an alternative, especially isolated as good as `esbuild` development server. 🎉 

I left some comments mostly on consistency and [planned extraction](https://github.com/sourcegraph/sourcegraph/issues/21905) of the development server logic into a separate workspace. The only blocking issue I noticed during local testing is related to CSS recompilation on file change: some CSS changes are not picked — the cached stylesheet is served instead. See the comment below in the `stylePlugin.ts` diff.

A couple of less important issues:

1. It seems that `autoprefixer` is not applied to _some_ stylesheets: 

ESBuild output:

<img width="439" alt="Screenshot 2021-09-01 at 10 10 00" src="https://user-images.githubusercontent.com/3846380/131655560-c91e83f2-3a6c-45b8-9f92-518ba1b603ac.png">

Webpack output:

<img width="435" alt="Screenshot 2021-09-01 at 13 31 23" src="https://user-images.githubusercontent.com/3846380/131656523-7eb43d33-9fef-4751-bbc0-a7e8767cabeb.png">

This issue causes some visual bugs:

<img width="592" alt="Screenshot 2021-09-01 at 10 10 12" src="https://user-images.githubusercontent.com/3846380/131655808-9ec3ff7c-ce66-40a7-b4ac-319eff779a94.png">

2. Clicking `Find references` on the `HoverOverlay` results in an uncaught exception: 

<img width="1246" alt="Screenshot 2021-09-01 at 13 29 03" src="https://user-images.githubusercontent.com/3846380/131656191-f29b151e-9682-49e3-83bd-4f655cc1e999.png">

Looking forward to getting this merged and trying it for day-to-day development!

eseliger reviewed (approved):
Nice 🚀 Love the performance. Thank you for investing so much time here!

Tested it briefly locally: Overall it works great. Some little things I found:
- yaml syntax highlighting in monaco is broken (json works)
- The icon issue that Valery already posted about
- For some reason image assets have a very high TTFB for me, around 1s.
- Something around styles must be off, the nav bar isn't responsive anymore

As a side-note, I think we should be cautious as long as we _ship_ webpack bundles to not have a too high adoption of esbuild in our dev team, so we still catch things that are broken/different in the webpack version. Also, as long as this isn't super stable, we should also encourage people to try webpack locally before filing a bug they find in their dev env.

sqs commented on client/web/dev/esbuild/stylePlugin.ts:
```diff
@@ -0,0 +1,124 @@
+    setup: build => {
+        const modulesMap = new Map<string, string>()
+        const modulesPlugin = postcssModules({
+            generateScopedName: '[name]__[local]', // omit hash for local dev
+            localsConvention: 'camelCase',
```
They're actually different config types (eg `exportLocalsConvention` vs `localsConvention`, `localIdentName` vs `generateScopedName`) because they use different underlying libs apparently. We could extract the string values, but the only one that needs to be the same is `camelCase`, so I don't think it's worth it.

sqs commented on doc/dev/background-information/web/build.md:
```diff
@@ -50,3 +50,29 @@ Run `yarn add PACKAGE` or `yarn add -D PACKAGE` in the root directory.
+
+- It's probably possible to configure Webpack to be faster and produce smaller dev bundles, so consider these comparisons as reflecting the current state, not the hypothetical ideal state after more optimization.
+- Webpack is still used for all other web builds (including storybooks and the browser extension).
+- esbuild is not configured to make a production build. Just use it for local dev for now.
```
It seems like it could go in either one. I'd like to avoid making this PR bigger by including an unrelated doc move in it, but feel free to move it after the fact.

sqs commented on doc/dev/background-information/web/build.md:
```diff
@@ -50,3 +50,29 @@ Run `yarn add PACKAGE` or `yarn add -D PACKAGE` in the root directory.
+Comparison vs. Webpack:
+
+- esbuild: faster initial build (esbuild ~3s vs. Webpack ~53s)
+- esbuild: faster recompilation (esbuild ~900ms vs. Webpack ~5000ms)
```
Yeah. Is it faster for you?

valerybugakov replied:
Yep, recompilation is around 2s on the first file change. It goes down to 1s if I make consequent changes to the same file.

sqs replied:
Hmm. For example, if I add an `x` literal text to the JSX in `UserProfileFormFields`, it takes ~5s each time to recompile. This is on a big Linux workstation. Maybe I've just been experiencing slower recompiles than everyone else all along, thus my particular desire for esbuild?

sqs commented on doc/dev/background-information/web/build.md:
```diff
@@ -50,3 +50,29 @@ Run `yarn add PACKAGE` or `yarn add -D PACKAGE` in the root directory.
 
 Run `yarn upgrade -L PACKAGE`.
+
+### esbuild
```
Added a link to the readme to https://docs.sourcegraph.com/dev/background-information/web/build#esbuild for visibility.

sqs commented on client/web/dev/esbuild/build.ts:
```diff
@@ -0,0 +1,114 @@
+import * as path from 'path'
+
+import * as esbuild from 'esbuild'
+import signale from 'signale'
```
esbuild has no default export, so the `*` is needed. (Oddly, `import esbuild from 'esbuild'` works to refer to types, but not actual exported bindings.)

You are right about `path`, though. Thanks!

sqs commented on client/web/dev/esbuild/build.ts:
```diff
@@ -0,0 +1,114 @@
+        },
+    ],
+    define: {
+        'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV || 'development'),
+        'process.env.PERCY_ON': JSON.stringify(process.env.PERCY_ON),
+        'process.env.SOURCEGRAPH_API_URL': JSON.stringify(process.env.SOURCEGRAPH_API_URL),
```
(mentioned in a separate comment) Not yet, but I will look into `sg` and standalone usage after merging this PR.

valerybugakov replied:
Sounds good, thanks!

sqs commented on client/web/dev/esbuild/stylePlugin.ts:
```diff
@@ -0,0 +1,124 @@
+            const cached = transformCache.get(inputPath)
+            if (cached && cached.inputContents === inputContents) {
+                return cached
+            }
```
Good catch, thanks!

The stylePlugin was not properly invalidating the cache for SourcegraphWebApp.scss when one of its included files changed. I fixed this by adding cache logic that also checks the included files. Confirmed working now.

valerybugakov commented on doc/dev/background-information/web/build.md:
````diff
@@ -50,3 +50,29 @@ Run `yarn add PACKAGE` or `yarn add -D PACKAGE` in the root directory.
+To use esbuild instead of Webpack, set the env var `DEV_WEB_BUILDER=esbuild`:
+
+``` shell
+DEV_WEB_BUILDER=esbuild enterprise/dev/start.sh
````
Got it 👍

sqs commented:
@eseliger:

Thanks for checking it out!

> Overall it works great. Some little things I found:
>
> yaml syntax highlighting in monaco is broken (json works)

Hmm, I fixed it by adding `yaml` as a language in the monaco-editor webpack plugin settings (which are shared with esbuild). It was omitted before. :shrug: 

> The icon issue that Valery already posted about

Added to https://github.com/sourcegraph/sourcegraph/issues/24518

> For some reason image assets have a very high TTFB for me, around 1s.

This is "by design" in esbuild's dev server. Instead of watching or polling, it actually executes a complete rebuild for each page load. The way this is implemented is that it caches a rebuild for [250ms](https://sourcegraph.com/github.com/evanw/esbuild@9f507f476cf4a620de0f56c5940ee94db0e40c5c/-/blob/pkg/api/serve_other.go?L62:1), which it assumes is long enough for all assets to load. This usually works, but it doesn't remain cached when navigating to other pages after the initial page load, so you need to wait ~1s for the full rebuild again. I think a different impl would be better for esbuild's dev server, but I understand why it is this way for now (it's simpler and reduces the risk of getting a stale asset).

> Something around styles must be off, the nav bar isn't responsive anymore

Hmm, I see it being responsive. Noted at https://github.com/sourcegraph/sourcegraph/issues/24518 and will follow up later.

sqs commented:
@valerybugakov Thank you for the review!

> It seems that autoprefixer is not applied to some stylesheets:

Fixed (my invocation of `postcss(...)` was incorrect). Unfortunately, this adds ~500ms to the initial build time (likely `autoprefixer` and other postcss plugins actually running on the big SCSS file).

sqs commented:
@valerybugakov:

> Clicking Find references on the HoverOverlay results in an uncaught exception

Fixed, thanks! It ended up being a problem whereby we were importing some `rxjs/internal/...` stuff, which caused a different (and `instanceof`-incompatible) variant of `rxjs` to be loaded alongside the usual one. Hacked it with 2 packageResolution entries (eg `'rxjs/internal/OuterSubscriber' -> require.resolve('rxjs/_esm5/internal/OuterSubscriber')`) for esbuild. I'm curious as to why this wasn't needed for Webpack and whether it's an esbuild bug or just a Webpack do-what-I-mean feature.
